### PR TITLE
LibCore: Refractor ArgsParser and add support for some Optional types

### PR DESCRIPTION
**Pull request by LucasChollet**
This refactoring remove a lot of duplicate code and also allows to easily add support for new types (at least some).

### CODE REVIEW

File: AK/Concepts.h [hunk 1]
```diff
@@ -111,6 +111,12 @@ concept IterableContainer = requires
     { declval<T>().begin() } -> IteratorPairWith<decltype(declval<T>().end())>;
 };
 
+template<typename T, typename InnerType = typename T::ValueType>
+inline constexpr bool IsAppendableContainer = requires(T container, InnerType element)
```
Comment (kleinesfilmroellchen) [inline AK/Concepts.h:115]: This is not a concept; if you're not adding a concept then why do you say so in the commit message?

File: Userland/Libraries/LibCore/ArgsParser.cpp [hunk 2]
```diff
@@ -568,143 +568,51 @@ void ArgsParser::add_positional_argument(Arg&& arg)
...
             return opt.has_value();
-        }
-    };
-    add_positional_argument(move(arg));
+        });
 }
 
-void ArgsParser::add_positional_argument(Vector<char const*>& values, char const* help_string, char const* name, Required required)
+template<bool IsContainer>
+void ArgsParser::add_positional_argument_helper(char const* help_string, char const* name, Required required, Function<bool(char const*)> accept_value)
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCore/ArgsParser.cpp:509]: The template argument should not be named like the "concept" to avoid confusion. We name non-type template arguments like variables.

Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibCore/ArgsParser.cpp:509]: Additionally, it looks like a bool-based enum would be much more readable here.

File: Userland/Libraries/LibCore/ArgsParser.h [hunk 3]
```diff
@@ -119,6 +126,9 @@ class ArgsParser {
     void add_positional_argument(double& value, char const* help_string, char const* name, Required required = Required::Yes);
 
 private:
+    template<bool RequireArgument = true>
+    void add_option_helper(char const* help_string, char const* long_name, char short_name, char const* value_name, OptionHideMode hide_mode, Function<bool(char const*)> accept_value);
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCore/ArgsParser.h:142]: Again, please (1) use normal variable naming for non-type template arguments, and (2) create a bool-based enum to make all the calls to this function more readable.

Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibCore/ArgsParser.h:142]: And the helper can easily receive Strings, no need to keep around old cruft if you're refactoring anyways :^). Now that I see that, please try to do that everywhere else as well, as far as it's possible with the many users.

File: Userland/Libraries/LibCore/ArgsParser.h [hunk 4]
```diff
@@ -121,6 +121,18 @@ class ArgsParser {
             });
     }
 
+    template<typename T>
+    requires(IsOneOf<T, Optional<String>, Optional<StringView>>) void add_positional_argument(T& value, char const* help_string, char const* name, Required required = Required::Yes)
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCore/ArgsParser.h:125]: `String`s here as well

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
Generally: If you're refactoring an API, you might as well remove old cruft like the c-strings we have all over ArgsParser. A total removal from public APIs though can be its own commit or PR, and you don't *have* to do that here. I'd love to see it though :^)

File: Userland/Libraries/LibCore/ArgsParser.cpp (hunk 2 above)
Comment (alimpfard) [reply Userland/Libraries/LibCore/ArgsParser.cpp:509]: I'll add that making that a normal bool argument won't hurt anything, or even passing whatever `INT_MAX` is from the caller

File: Userland/Libraries/LibCore/ArgsParser.cpp [hunk 5]
```diff
@@ -568,143 +568,51 @@ void ArgsParser::add_positional_argument(Arg&& arg)
...
-        INT_MAX,
-        [&values](char const* s) {
-            values.append(s);
-            return true;
-        }
-    };
-    add_positional_argument(move(arg));
-}
+template void ArgsParser::add_positional_argument_helper<true>(char const*, char const*, Required, Function<bool(char const*)>);
+template void ArgsParser::add_positional_argument_helper<false>(char const*, char const*, Required, Function<bool(char const*)>);
```
Comment (alimpfard) [inline Userland/Libraries/LibCore/ArgsParser.cpp:522]: Why are these here? they shouldn't be needed here, all the instances are in the cpp file.

File: Userland/Libraries/LibCore/ArgsParser.h [hunk 6]
```diff
@@ -6,10 +6,12 @@
 
 #pragma once
 
+#include <AK/Concepts.h>
 #include <AK/Function.h>
 #include <AK/String.h>
 #include <AK/Vector.h>
 #include <LibMain/Main.h>
+#include <climits>
```
Comment (alimpfard) [inline Userland/Libraries/LibCore/ArgsParser.h:14]: Nope.
Use `limits.h`.

File: Userland/Libraries/LibCore/ArgsParser.h [hunk 7]
```diff
@@ -91,17 +93,35 @@ class ArgsParser {
...
+    {
+        add_positional_argument_helper<true>(help_string, name, required,
+            [&values](char const* s) {
+                values.append(s);
+                return true;
+            });
+    }
+
+    template<typename T>
+    requires(IsOneOf<T, char const*, String, StringView>) void add_positional_argument(T& value, char const* help_string, char const* name, Required required = Required::Yes)
```
Comment (alimpfard) [inline Userland/Libraries/LibCore/ArgsParser.h:115]: You can collapse this down to
```suggestion
    template<OneOf<char const*, String, StringView> T>
    void add_positional_argument(T& value, char const* help_string, char const* name, Required required = Required::Yes)
```

File: Userland/Libraries/LibCore/ArgsParser.cpp [hunk 8]
```diff
@@ -556,13 +444,32 @@ void ArgsParser::add_option(Vector<size_t>& values, char const* help_string, cha
...
-        },
+        });
+}
+
+template void ArgsParser::add_option(char const*&, char const*, char const*, char, char const*, OptionHideMode);
+template void ArgsParser::add_option(StringView&, char const*, char const*, char, char const*, OptionHideMode);
+template void ArgsParser::add_option(String&, char const*, char const*, char, char const*, OptionHideMode);
+
+template<bool RequireArgument>
+void ArgsParser::add_option_helper(char const* help_string, char const* long_name, char short_name, char const* value_name, ArgsParser::OptionHideMode hide_mode, Function<bool(char const*)> accept_value)
```
Comment (alimpfard) [inline Userland/Libraries/LibCore/ArgsParser.cpp:455]: Yeah no, please just make the template argument a normal one, it's really not worth generating two instances and then merging them just to pass a single bool in a very cold function.

File: Userland/Libraries/LibCore/ArgsParser.h [hunk 9]
```diff
@@ -82,16 +82,23 @@ class ArgsParser {
...
     void add_option(unsigned& value, char const* help_string, char const* long_name, char short_name, char const* value_name, OptionHideMode hide_mode = OptionHideMode::None);
     void add_option(double& value, char const* help_string, char const* long_name, char short_name, char const* value_name, OptionHideMode hide_mode = OptionHideMode::None);
     void add_option(Optional<double>& value, char const* help_string, char const* long_name, char short_name, char const* value_name, OptionHideMode hide_mode = OptionHideMode::None);
     void add_option(Optional<size_t>& value, char const* help_string, char const* long_name, char short_name, char const* value_name, OptionHideMode hide_mode = OptionHideMode::None);
     void add_option(Vector<size_t>& values, char const* help_string, char const* long_name, char short_name, char const* value_name, char separator = ',', OptionHideMode hide_mode = OptionHideMode::None);
 
+    template<typename T>
+    requires(IsOneOf<T, char const*, String, StringView>) void add_option(T& value, char const* help_string, char const* long_name, char short_name, char const* value_name, OptionHideMode hide_mode = OptionHideMode::None)
```
Comment (alimpfard) [inline Userland/Libraries/LibCore/ArgsParser.h:93]: Same thing with `template<OneOf<...> T>` (and more in the other commits)

**Review by alimpfard — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibCore/ArgsParser.h (hunk 6 above)
Comment (LucasChollet) [reply Userland/Libraries/LibCore/ArgsParser.h:14]: What is the reason behind this?

File: Userland/Libraries/LibCore/ArgsParser.cpp (hunk 5 above)
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibCore/ArgsParser.cpp:522]: My comment on this got lost somehow...

File: Userland/Libraries/LibCore/ArgsParser.h (hunk 7 above)
Comment (LucasChollet) [reply Userland/Libraries/LibCore/ArgsParser.h:115]: Very neat!

File: Userland/Libraries/LibCore/ArgsParser.h (hunk 4 above)
Comment (LucasChollet) [reply Userland/Libraries/LibCore/ArgsParser.h:125]: Wouldn't a `StringView` be better?

**Comment by LucasChollet** [comment]
I was worried that removing `char const*` from `ArgsParser` was too many consequences and changes for something not very critical like this. As kleinesfilmroellchen seems to be ok with it, I will add those changes to this PR.

File: Userland/Libraries/LibCore/ArgsParser.cpp (hunk 5 above)
Comment (LucasChollet) [reply Userland/Libraries/LibCore/ArgsParser.cpp:522]: I agree with you, and I can't tell you why, but is doesn't compile without it.

Either way, I will remove those when making the compile-time argument a runtime one.

File: Userland/Libraries/LibCore/ArgsParser.cpp (hunk 8 above)
Comment (LucasChollet) [reply Userland/Libraries/LibCore/ArgsParser.cpp:455]: I did it because to indicate to the reader that they are compile-time argument, but wasn't really sure of the design and really wanted a review haha.
